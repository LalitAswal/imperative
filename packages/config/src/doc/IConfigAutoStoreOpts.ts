/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { ICommandArguments } from "../../../cmd/src/doc/args/ICommandArguments";
import { IHandlerParameters } from "../../../cmd/src/doc/handler/IHandlerParameters";

export interface IConfigAutoStoreFindActiveProfileOpts {
    params?: IHandlerParameters;
    profileProps?: string[];
    profileTypes?: string[];
    defaultProfileName?: string;
}

export interface IConfigAutoStoreFindAuthHandlerForProfileOpts extends IConfigAutoStoreFindActiveProfileOpts {
    profilePath?: string;
    cmdArguments?: ICommandArguments;
    defaultBaseProfileName?: string;
}

export interface IConfigAutoStoreStoreSessCfgPropsOpts extends IConfigAutoStoreFindAuthHandlerForProfileOpts {
    sessCfg?: { [key: string]: any };
    propsToStore?: string[];
    profileName?: string;
    profileType?: string;
}

