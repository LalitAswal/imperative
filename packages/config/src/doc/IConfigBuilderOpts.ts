/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { IConfig } from "./IConfig";
import { IProfileProperty } from "../../../profiles";

export interface IConfigBuilderOpts {
    /**
     * Specify true to populate default values for profile properties with the
     * `IProfileProperty.includeInTemplate` flag set to true.
     */
    populateProperties?: boolean;

    /**
     * Existing config properties that the new config will be merged with. If
     * set, the config builder skips prompting for properties that already have
     * values defined.
     */
    mergeConfig?: IConfig;

    /**
     * Callback that prompts the user to enter a value for a profile property.
     * The method will be called to populate missing values in the base profile.
     * It should return the value that the user entered, or null or undefined if the prompt was cancelled.
     * @param propName The name of the property
     * @param property The profile property definition
     */
    getValueBack? (propName: string, property: IProfileProperty): Promise<any>;
}
