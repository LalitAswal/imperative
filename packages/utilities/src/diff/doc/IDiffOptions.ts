type outputFormat = "html" | "unifiedstring" | "terminal";

/**
 * interface for diff options
 */
export interface IDiffOptions {

    /**
     * Output format of differences between two, to be returned
     * @type {outputFormat}
     * @memberOf IDiffOptions
     */
    outputFormat: outputFormat,

    /**
     * Number of context line arguments
     * @type {number}
     * @memberOf IDiffOptions
     */
    contextLinesArg?: number,

}