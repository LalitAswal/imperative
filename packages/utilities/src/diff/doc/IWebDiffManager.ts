/**
 * Web help manager API that handles launching of web help and generating it if necessary.
 * @export
 * @interface IWebDiffManager
 */
export interface IWebDiffManager {
    /**
     * launch the diff of two file in web
     * @param {string} content Html content string to launch in web
     */
    openDiffs(content: string): void;

}
