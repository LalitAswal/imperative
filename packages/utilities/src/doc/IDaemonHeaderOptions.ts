
/**
 * Option interface to construct headers for daemon mode protocol
 * @export
 * @interface IDaemonHeaderOptions
 */
export interface IDaemonHeaderOptions {

    /**
     * Process exit code
     * @type {number}
     * @memberof IDaemonHeaderOptions
     */
    exitCode?: number;

    /**
     * Indicator for prompting
     * @type {boolean}
     * @memberof IDaemonHeaderOptions
     */
    prompt?: boolean;

    /**
     * Content is for stdout
     * @type {boolean}
     * @memberof IDaemonHeaderOptions
     */
    stdout?: boolean;

    /**
     * Content is for stderr
     * @type {boolean}
     * @memberof IDaemonHeaderOptions
     */
    stderr?: boolean;
}