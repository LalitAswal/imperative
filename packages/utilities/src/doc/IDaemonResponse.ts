/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/


/**
 * Option interface to construct response from daemon client
 * @export
 * @interface IDaemonResponse
 */
export interface IDaemonResponse {

    /**
     * Content is from daemon client, not an interactive user
     * @type {string}
     * @memberof IDaemonResponse
     */
    id: string;

    /**
     * Content is reply from daemon
     * @type {string}
     * @memberof IDaemonResponse
     */
    reply?: string;
}