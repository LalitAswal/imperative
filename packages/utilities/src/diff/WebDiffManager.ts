/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import * as fs from "fs";
import * as path from "path";

import { Constants } from "../../../constants/src/Constants";
import {  ProcessUtils, GuiResult } from "../../../utilities/src/ProcessUtils";
import { ImperativeConfig } from "../../../utilities/src/ImperativeConfig";
import WebDiffGenerator from "./WebDiffGenerator";
import { IWebDiffManager } from "./doc/IWebDiffManager";
import { ImperativeError } from "../../../error";
import { html } from "diff2html";

/**
 * Imperatice Web Differences Manager handles the opeing of diffs and
 * constructs the dirs and files if necessary
 * @export
 * @class WebDiffManager
 */
export class WebDiffManager implements IWebDiffManager {
    /**
     * Singleton instance of this class
     * @private
     * @static
     * @type {WebHelpManager}
     * @memberof WebHelpManager
     */
    private static mInstance: WebDiffManager = null;

    /**
     * Return a singleton instance of this class
     * @static
     * @readonly
     */
    public static get instance(): WebDiffManager {
        if (this.mInstance == null) {
            this.mInstance = new WebDiffManager();
        }

        return this.mInstance;
    }


    /**
     * Launch help page for specific group/command in browser.
     * @param {string} inContext - content of diff to be shown
     * @param {IHandlerResponseApi} cmdResponse - Command response object to use for output
     * @memberof WebHelpManager
     */
    public async openDiffs(patchDiff: string) {
        const doWeHaveGui = ProcessUtils.isGuiAvailable();
        if (doWeHaveGui !== GuiResult.GUI_AVAILABLE) {
            let errMsg = "You are running in an environment with no graphical interface." +
                "\nAlternatively, you can run '" + ImperativeConfig.instance.findPackageBinName() +
                " --help' for text-based help.";
            if (doWeHaveGui === GuiResult.NO_GUI_NO_DISPLAY) {
                errMsg += "\n\nIf you are running in an X Window environment," +
                    "\nensure that your DISPLAY environment variable is set." +
                    "\nFor example, type the following:" +
                    "\n    echo $DISPLAY" +
                    "\nIf it is not set, assign a valid value. For example:" +
                    "\n    export DISPLAY=:0.0" +
                    "\nThen try the --help-web option again.";

                throw new ImperativeError({
                    msg: errMsg
                });
            }

            return;
        }

        if (this.checkWebDiffDirExists()) await new WebDiffGenerator(ImperativeConfig.instance, this.webDiffDir).buildDiffDir();

        const htmlDiff = await html(patchDiff, {
            outputFormat: "side-by-side",
            matching: "lines",
            diffStyle: "char",
        });

        if (htmlDiff != null) {
            // writing the diff content into a textt file
            fs.writeFileSync(path.join(this.webDiffDir, 'index.html'), this.genHtmlForDiffs(htmlDiff, patchDiff));
            try {
                ProcessUtils.openInDefaultApp(`file://${this.webDiffDir}/index.html`);
            } catch (e) {
                throw new ImperativeError({
                    msg: "Failed to launch web diff, try running -h for console help instead",
                    causeErrors: [e]
                });
            }
        }
    }


    /**
     * Check if the web diff direcory base has been generated
     * at cli home
     */
    private async checkWebDiffDirExists() {
        if (fs.existsSync(this.webDiffDir)) {
            return true;
        }
        return false;
    }


    /**
     * Gets the directory where built copy of web diff launcher is stored
     * @readonly
     * @private
     * @returns {string} Absolute path of directory
     */
    private get webDiffDir(): string {
        return path.join(ImperativeConfig.instance.cliHome, Constants.WEB_DIFF_DIR);
    }

    /**
     * Returns header HTML for help page
     * @private
     * @param title - Title string for the page
     */
    private genHtmlForDiffs(htmlDiff: string, unifiedStringDiff: string): string {
        return `<!DOCTYPE html>
        <html>
          <head>
            <link
              rel="stylesheet"
              href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.7.1/styles/github.min.css"
            />
            <link
              rel="stylesheet"
              type="text/css"
              href="https://cdn.jsdelivr.net/npm/diff2html/bundles/css/diff2html.min.css"
            />
            <script
              type="text/javascript"
              src="https://cdn.jsdelivr.net/npm/diff2html/bundles/js/diff2html-ui.min.js"
            ></script>
        
            <script>
                const fr = new FileReader()
        
                fr.onload(()=>{
                    document.getElementById('diffOutput').textContent = fr.result
                })
        
                fr.readAsText()
            const diffString = ${unifiedStringDiff}
        
              document.addEventListener('DOMContentLoaded', function () {
                var targetElement = document.getElementsByClassName('d2h-file-list-wrapper')[0];
                var configuration = {
                  drawFileList: true,
                  fileListToggle: false,
                  fileListStartVisible: false,
                  fileContentToggle: false,
                  matching: 'lines',
                  outputFormat: 'side-by-side',
                  synchronisedScroll: true,
                  highlight: true,
                  renderNothingWhenEmpty: false,
                };
                var diff2htmlUi = new Diff2HtmlUI(targetElement, diffString, configuration);
                diff2htmlUi.draw();
                diff2htmlUi.highlightCode();
              });
            </script>
        
            <meta content="0; url=diff.html?p=" />
          </head>
          <body>
            ${htmlDiff}
          </body>
        </html>
    `;
    }
}
