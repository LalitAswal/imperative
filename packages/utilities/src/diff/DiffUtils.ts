/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

/* eslint-disable no-console */
import { diff } from "jest-diff";
import { createTwoFilesPatch } from "diff";
import { IDiffOptions } from "./doc/IDiffOptions";
import { TextUtils } from "../TextUtils";
import { html } from "diff2html";
import { WebDiffManager } from "./WebDiffManager";

/**
 * Utilities to get the diff strings and open the diff strings in terminal and in browser
 * @export
 * @class DiffUtils
 */
export class DiffUtils {
    /**
     * Get the differnce between two string in the form of html, unifiedString and terminal output depending uponn the
     * options passed into the funtions
     * @param {string} string1
     * @param {string} string2
     * @param {string} options
     * @returns {Promise<string>}
     */
    public static async getDiffString(string1: string, string2: string, options: IDiffOptions): Promise<string> {

        if (options.outputFormat === 'terminal') {
            let expandflag = true;
            if (options.contextLinesArg >= 0) {
                expandflag = false;
            }
            const jsonDiff = await diff(string1, string2, {
                aAnnotation: "Removed",
                bAnnotation: "Added",
                aColor: TextUtils.chalk.red,
                bColor: TextUtils.chalk.green,
                contextLines: options.contextLinesArg,
                expand: expandflag
            });

            return jsonDiff;
        }

        if (options.outputFormat === 'unifiedstring' || options.outputFormat === "html") {
            const patchDiff = createTwoFilesPatch(
                'file-a', 'file-b', string1, string2
            );

            if (options.outputFormat === 'html') {
                const htmlDiff = await html(patchDiff, {
                    outputFormat: "side-by-side",
                    matching: "lines",
                    diffStyle: "char",
                });

                return htmlDiff;
            }
            return patchDiff;
        }
    }

    /**
 * Get the differnce between two string in browser
 * @param {string} string1
 * @param {string} string2
 * @return {void}
 */
    public static async openDiffInbrowser(string1: string, string2: string) {
        const patchDiff = createTwoFilesPatch(
            'file-a', 'file-b', string1, string2
        );

        await WebDiffManager.instance.openDiffs(patchDiff);

    }
}