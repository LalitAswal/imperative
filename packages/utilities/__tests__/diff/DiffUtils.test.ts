/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { DiffUtils } from "../../src/diff/DiffUtils";
import { IDiffOptions } from "../../src/diff/doc/IDiffOptions";
import { WebDiffManager } from "../../src/diff/WebDiffManager";

describe("DiffUtils", () => {

    describe("getDiffString", () => {
        let string1: string;
        let string2: string;
        let options: IDiffOptions;

        it("should return a diff string", async () => {
            string1 = "random string one";
            string2 = "random string two";
            options = { outputFormat: "unifiedstring" };
            expect(await DiffUtils.getDiffString(string1, string2, options)).toMatchSnapshot();

        });
    });

    describe("openDiffInbrowser", () => {

        it("should open the diffs in browser", async () => {
            const createTwoFilesPatchSpy = jest.fn().mockReturnValue("test");
            jest.doMock("diff", () => {
                return {
                    ...(jest.requireActual('diff')),
                    createTwoFilesPatch: createTwoFilesPatchSpy
                };
            });
            const string1 = "test string one";
            const string2 = "test string two";

            const openDiffSpy = jest.spyOn(WebDiffManager.instance, "openDiffs").mockImplementation(jest.fn());
            await DiffUtils.openDiffInbrowser(string1, string2);
            expect(createTwoFilesPatchSpy).toHaveBeenCalledWith('file-a', 'file-b', string1, string2);
            expect(openDiffSpy).toHaveBeenCalledWith("test");
        });
    });

});
