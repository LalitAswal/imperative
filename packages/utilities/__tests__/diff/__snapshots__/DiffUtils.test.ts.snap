// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiffUtils getDiffString should return a diff string 1`] = `
"===================================================================
--- file-a
+++ file-b
@@ -1,1 +1,1 @@
-random string one
\\\\ No newline at end of file
+random string two
\\\\ No newline at end of file
"
`;
